src/bentoml/_internal/io_descriptors/json.py

```diff
@@ -19,11 +20,29 @@
 from ...exceptions import InvalidArgument
 from ..service.openapi import REF_PREFIX
 from ..service.openapi import SUCCESS_DESCRIPTION
+from ..service.openapi.utils import typed_dict_to_dict
 from ..service.openapi.specification import Schema
 from ..service.openapi.specification import MediaType
 
 EXC_MSG = "'pydantic' must be installed to use 'pydantic_model'. Install with 'pip install bentoml[io-json]'."
 
+if sys.version_info >= (3, 11):
+    from typing import TypedDict
```

Comment:
is there a change to typeddict from 3.11 onwards :(

Reply:
In 3.11 and below, TypedDict seems to have few exception issues to consider. 

As you can see from pydantic's implementation of TypedDict, they have complicated exception handling for this issue. 
https://github.com/pydantic/pydantic/pull/2216/files#diff-bde859c4d65789b57bf14f734ec9ebabc83b30a82f6b0765a650ac18a25f0950R15
https://github.com/pydantic/pydantic/blob/v1.10.7/pydantic/annotated_types.py#L22

In implementing the TypedDict feature in this PR, I experienced a number of issues so did not use the internal features of TypedDict ( like a `__required_keys__`, `there is no way to reflect Required/NotRequired keys.`) which are the reason why Pydantic wanted to exception handle the availability of typing.TypedDict for each version(<=3.8 ,  >=3.11). 
So `sys.version_info >= (3, 11)` may not be necessary at this point. 
However, without this condition, someone trying to modify this logic would have to take into account the subtle differences in the internal implementation of TypedDict for each python version.

In fact, the exception handling below is also unnecessary at this point. 
~~~Python
if is_legacy_typeddict(typeddict):
            raise BadInput(
                "You should use `typing_extensions.TypedDict` instead of `typing.TypedDict` with Python < 3.11."
            )
~~~

it is fine to use.
~~~Python
from typing import TypedDcit
from typing_extension import NotRequried 

from typing import List, Union, TypedDict


from typing_extensions import NotRequired

class IrisFeatures(TypedDict, total=False): # it's working 
    sepal_len: Union[float, int]
    sepal_width: float
    petal_len: Union[float , int]
    petal_width: float

    request_id: NotRequired[int]


JSON(typeddict=IrisFeatures)
~~~
but i'm not sure some side effect

If we're just using TypedDict for our Open API Schema implementation like the current implementation,
"sys.version_info >= (3, 11)" is unnecessary. 

Is it appropriate to remove  that `sys.version_info >= (3, 11)` ?

Reply:
if i understand correct, the `typing_extension` should only act as backward compat for newer feature in older python version. So I think if we want to support TypedDict, then we will have to support similar to what pydantic did.


```python
from typing import List, Union, TypedDict


from typing_extensions import NotRequired
```

> However, without this condition, someone trying to modify this logic would have to take into account the subtle differences in the internal implementation of TypedDict for each python version.

Not sure if I understand this, but I'm surprised that they introduced breaking change among TypedDict iteration of Python versions.

I don't think TypedDict can be imported here in 3.7
Also in the code you can add in line documentation that reference back to pydantic's source code (credit is due where it should be)

P/s: I just read a bit more about TypedDict, it does seems they are slightly different among Python version 😮‍💨 

p/s: +1 that we probably need to do something similar to Pydantic

Reply:
> So I think if we want to support TypedDict, then we will have to support similar to what pydantic did.

If we want to support the same as pydantic, we may not need to have complex constraints like pydantic at this point

because pydantic is not only used OAS3 Schema but also Validation with type hint so they need to deeply each TypedDict feature. 

for example pydantic should raise Validation Error  in below example
https://docs.pydantic.dev/latest/usage/types/#typeddict
~~~Python
class AA(TypedDict)
      a: int
      b: str

class BB(BaseModel):
     aa: AA


BB(aa ={"a": "1", b: 1000}) # raise validation error
~~~

but this pullrequest is use TypedDict for type hint



56e3763 (here is (`>=3.11` condition ) is removed version)  
it is safe , if we only use to  OpenAPI Schema generating